py/optimize.py

```diff
@@ -0,0 +1,60 @@
+from scipy.optimize import differential_evolution
+
+def optimize(input_dict, loss_metric):
+
+  '''
+  Arguments:
+  - input_dict = a dict with format {category: (min, max)} specifying the bounds for UBI amounts
+                 for each category. If min == max, the amount is fixed.
+  - loss_metric = the type of loss metric to be used in optimization.
+
+  Returns:
+  - prints the loss and corresponding UBI amount per category solution set for each function evaluation.
+  - return an OptimizeResult with the optimal solution.
+  '''
+  
+  # Declare categories 
+  categories = ['senior', 'adult', 'child', 'dis_1', 'dis_2', 'dis_3','NORTH_EAST', 'NORTH_WEST', 
+                'YORKSHIRE', 'EAST_MIDLANDS', 'WEST_MIDLANDS', 'EAST_OF_ENGLAND', 'LONDON', 
+                'SOUTH_EAST', 'SOUTH_WEST', 'WALES', 'SCOTLAND', 'NORTHERN IRELAND']
+
+  # Assign values for ordering
+  senior = input_dict['senior']
+  child = input_dict['child']
+  dis_1 = input_dict['dis_1']
+  dis_2 = input_dict['dis_2']
+  dis_3 = input_dict['dis_3']
+  NORTH_EAST = input_dict['NORTH_EAST']
+  NORTH_WEST = input_dict['NORTH_WEST']
+  YORKSHIRE = input_dict['YORKSHIRE']
+  EAST_MIDLANDS = input_dict['EAST_MIDLANDS']
+  WEST_MIDLANDS = input_dict['WEST_MIDLANDS']
+  EAST_OF_ENGLAND = input_dict['EAST_OF_ENGLAND']
+  LONDON = input_dict['LONDON']
+  SOUTH_EAST = input_dict['SOUTH_EAST']
+  SOUTH_WEST = input_dict['SOUTH_WEST']
+  WALES = input_dict['WALES']
+  SCOTLAND = input_dict['SCOTLAND']
+  NORTHER_IRELAND = input_dict['NORTHERN_IRELAND']
+
+  # Declare bounds for each amount
+  bounds = [senior, child, dis_1, dis_2, dis_3, NORTH_EAST, NORTH_WEST, YORKSHIRE, EAST_MIDLANDS,

```

Comment:
could this instead be this? could skip lines 22-38 if so
```
bounds = [input_dict[i] for i in CATEGORIES]
```

Reply:
The bounds have to be in the same order as x, and since there's no guarantee the order of keys in input_dict will match that of CATEGORIES, I thought I needed lines 22-38 to reorder the variables into the specific order needed for x. Even if you make CATEGORIES = list(input_dict.keys()), I think you would need to reorder the keys to match the order of x before you can assign the variables using bounds = [input_dict[i] for i in CATEGORIES].

Reply:
I don't follow, this seems to work:
```
In [1]: input_dict = {'adult': (0, 200), 'child': (0, 150), 'senior': (0, 250)}

In [2]: CATEGORIES = ['senior', 'adult', 'child']

In [3]: bounds = [input_dict[i] for i in CATEGORIES]

In [4]: bounds
Out[4]: [(0, 250), (0, 200), (0, 150)]
```

Reply:
Sorry, I misunderstood - I'll add this in now.

Reply:
Sorry, I misunderstood - I'll add this in now.